### PR TITLE
AUTH-89: add audit options to oauth-server on oauth-audit-profile Variant03

### PR DESCRIPTION
**Pull request by ibihim**
Compare https://github.com/openshift/cluster-authentication-operator/pull/559

### CODE REVIEW

**Comment by ibihim** [comment]
@s-urbaniak , @stlaz , I have also a commit, that adds those code changes to the oauth-apiserver serverArguments, but the tests are failing in 2 places, which I would like to discuss with you.

In one place we have no additional bash-escaping changing a regex from ` --cors-allowed-origins=//127\.0\.0\.1(:|$) ` to `--cors-allowed-origins='//127\.0\.0\.1(:|$)' `. And in another place a hash changed...

**Comment by ibihim** [comment]
TODO:
- [ ] Verify tests test turn-off.
- [ ] Change test data for refactoring of oauth-apiserver server arguments.
- [ ] Try runAsUser 0 without privileged.

File: pkg/controllers/configobservation/oauth/observe_audit.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+	observedConfig := map[string]interface{}{}
+	observedAuditProfile := apiServer.Spec.Audit.Profile
+	if observedAuditProfile != configv1.NoneAuditProfileType {
+		if err := unstructured.SetNestedField(
+			observedConfig,
+			auditOptionsArgs,
+			ServerArgumentsPath...,
+		); err != nil {
+			return existingConfig, append(errs, fmt.Errorf(
```
Comment (s-urbaniak) [inline pkg/controllers/configobservation/oauth/observe_audit.go:62]: here, you use `append` while in line 48 you return `[]error{...}`. let's make this consistent.

**Comment by s-urbaniak** [comment]
this is great work @ibihim! 👏 I know it was a long road and complexities to understand, so really looking forward to have this finished soon! 🎉

File: pkg/controllers/deployment/default_deployment.go [hunk 2]
```diff
@@ -84,12 +85,32 @@ func getOAuthServerDeployment(
 	templateSpec.Volumes = append(templateSpec.Volumes, v...)
 	container.VolumeMounts = append(container.VolumeMounts, m...)
 
+	argsRaw, err := GetOAuthServerArgumentsRaw(&operatorConfig.Spec.ObservedConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get audit sync data: %w", err)
```
[the remark this answers is not in the document]
Comment (ibihim) [reply pkg/controllers/deployment/default_deployment.go:90]: I will lowercase all _Unable_.

File: bindata/oauth-openshift/deployment.yaml [hunk 3]
```diff
@@ -63,15 +63,21 @@ spec:
...
+              --v=${LOG_LEVEL} \
+              ${SERVER_ARGUMENTS}
           ports:
             - name: https
               containerPort: 6443
               protocol: TCP
           securityContext:
+            privileged: true
             readOnlyRootFilesystem: false # because of the `cp` in args
             runAsUser: 0 # because /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem is only writable by root
```
Comment (ibihim) [inline bindata/oauth-openshift/deployment.yaml:75]: @stlaz, @s-urbaniak, `runAsUser: 0` is already set. As I tested the `privileged: false` value before, I wouldn't do it again.

File: pkg/controllers/configobservation/oauth/observe_audit.go (hunk 1 above)
Comment (ibihim) [reply pkg/controllers/configobservation/oauth/observe_audit.go:66]: this was a nice comment. it makes the code cooler.

File: bindata/oauth-openshift/deployment.yaml (hunk 3 above)
Comment (s-urbaniak) [reply bindata/oauth-openshift/deployment.yaml:75]: sounds good then 👍

File: pkg/controllers/deployment/default_deployment.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibihim) [reply pkg/controllers/deployment/default_deployment.go]: @s-urbaniak, is this what you wanted?! :D [commit](https://github.com/ibihim/cluster-authentication-operator/commit/a7fe702a24bb14ea649d519f3360e94339ac275d)

File: pkg/controllers/deployment/default_deployment.go [hunk 4]
```diff
@@ -73,23 +74,43 @@ func getOAuthServerDeployment(
...
 	if err != nil {
-		return nil, fmt.Errorf("Unable to transform observed IDP sync data to volumes and mounts: %v", err)
+		return nil, fmt.Errorf("unable to transform observed IDP sync data to volumes and mounts: %v", err)
 	}
 	templateSpec.Volumes = append(templateSpec.Volumes, v...)
 	container.VolumeMounts = append(container.VolumeMounts, m...)
 
+	argsRaw, err := GetOAuthServerArgumentsRaw(&operatorConfig.Spec.ObservedConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get audit sync data: %w", err)
```
Comment (stlaz) [inline pkg/controllers/deployment/default_deployment.go:90]: `"unable to retrieve server arguments from observed config: %w", err`

File: pkg/controllers/deployment/default_deployment.go [hunk 5]
```diff
@@ -132,3 +153,26 @@ func appendEnvVar(envVars []corev1.EnvVar, envName, envVal string) []corev1.EnvV
 	}
 	return envVars
 }
+
+func GetOAuthServerArgumentsRaw(operatorConfig *runtime.RawExtension) (map[string]interface{}, error) {
+	oauthServerObservedConfig, err := common.UnstructuredConfigFrom(
```
Comment (stlaz) [inline pkg/controllers/deployment/default_deployment.go:158]: retrieve the unstructured config and pass it to both `GetOAuthServerArgumentsRaw()` (why is it exported?) and `GetOAuthServerArgumentsRaw()`

**Review by stlaz — COMMENTED** [review]
small things only

File: bindata/oauth-openshift/deployment.yaml [hunk 6]
```diff
@@ -63,15 +63,21 @@ spec:
...
               --config=/var/config/system/configmaps/v4-0-config-system-cliconfig/v4-0-config-system-cliconfig \
-              --v=${LOG_LEVEL}
+              --v=${LOG_LEVEL} \
+              ${SERVER_ARGUMENTS}
           ports:
             - name: https
               containerPort: 6443
               protocol: TCP
           securityContext:
+            privileged: true
```
[the remark this answers is not in the document]
Comment (ibihim) [reply bindata/oauth-openshift/deployment.yaml:73]: two lines below...

[the remark this answers is not in the document]
Comment (stlaz) [reply bindata/oauth-openshift/deployment.yaml:73]: well this is awkward

[the remark this answers is not in the document]
Comment (ibihim) [reply bindata/oauth-openshift/deployment.yaml:73]: https://github.com/openshift/cluster-authentication-operator/pull/563/files/0b10fd3704a36570b573c27e331391f60aa7aa7a..9e7cde854183bfa0908412c7503182d29d7c4abd#diff-9a542c37e688fe51096ad17db28f58240c8e732af38ec7f5192c46bf613aa13dR75

File: pkg/controllers/deployment/default_deployment.go (hunk 5 above)
Comment (ibihim) [reply pkg/controllers/deployment/default_deployment.go:158]: I moved the package, but forgot to unexport. Done.

File: pkg/controllers/common/arguments.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// ServerArguments type.
+func Parse(raw map[string]interface{}) (ServerArguments, error) {
+	args := make(ServerArguments)
+
+	for argName, argValue := range raw {
+		var argsSlice []string
+
+		argsSlice, found, err := unstructured.NestedStringSlice(raw, argName)
+		if !found || err != nil {
+			str, found, err := unstructured.NestedString(raw, argName)
```
[the remark this answers is not in the document]
Comment (ibihim) [reply pkg/controllers/common/arguments.go]: You are right, but I didn't write the code in the first place. It already existed like so and when removing `string` option, tests fail.

**Pull request merged**